File: cpp-exercises/02_temperature_converter.cpp
#include <cctype>
#include <iostream>
#include <stdexcept>
#include <string>

double parse_number(const std::string& text) {
    std::size_t processed = 0;
    const double value = std::stod(text, &processed);

    if (processed != text.size()) {
        throw std::invalid_argument("Temperature must be a number.");
    }

    return value;
}

char parse_unit(const std::string& text) {
    if (text.size() != 1) {
        throw std::invalid_argument("Unit must be C or F.");
    }

    const char unit = static_cast<char>(std::toupper(static_cast<unsigned char>(text[0])));

    if (unit != 'C' && unit != 'F') {
        throw std::invalid_argument("Unit must be C or F.");
    }

    return unit;
}

double celsius_to_fahrenheit(double celsius) {
    return celsius * 9.0 / 5.0 + 32.0;
}

double fahrenheit_to_celsius(double fahrenheit) {
    return (fahrenheit - 32.0) * 5.0 / 9.0;
}

int main() {
    std::string value_text;
    std::string unit_text;

    std::cout << "Temperature converter\n";
    std::cout << "Enter the temperature value: ";
    std::getline(std::cin, value_text);

    std::cout << "Is the value in C or F? ";
    std::getline(std::cin, unit_text);

    try {
        const double value = parse_number(value_text);
        const char unit = parse_unit(unit_text);

        std::cout << std::fixed;
        std::cout.precision(2);

        if (unit == 'C') {
            std::cout << value << " C is " << celsius_to_fahrenheit(value) << " F\n";
        } else {
            std::cout << value << " F is " << fahrenheit_to_celsius(value) << " C\n";
        }
    } catch (const std::exception& error) {
        std::cout << "Invalid input: " << error.what() << '\n';
        return 1;
    }

    return 0;
}

File: python-exercises/02_temperature_converter.py
def celsius_to_fahrenheit(celsius: float) -> float:
    return celsius * 9 / 5 + 32


def fahrenheit_to_celsius(fahrenheit: float) -> float:
    return (fahrenheit - 32) * 5 / 9


def parse_unit(text: str) -> str:
    unit = text.strip().upper()
    if unit not in {"C", "F"}:
        raise ValueError("Unit must be C or F.")
    return unit


def main() -> None:
    print("Temperature converter")

    try:
        value = float(input("Enter the temperature value: ").strip())
        unit = parse_unit(input("Is the value in C or F? ").strip())
    except ValueError as error:
        print(f"Invalid input: {error}")
        return

    if unit == "C":
        converted = celsius_to_fahrenheit(value)
        print(f"{value:.2f} C is {converted:.2f} F")
    else:
        converted = fahrenheit_to_celsius(value)
        print(f"{value:.2f} F is {converted:.2f} C")


if __name__ == "__main__":
    main()
